kv/kvserver: TestMergeQueue failed

kv/kvserver.TestMergeQueue [failed](https://tanzanite.cluster.engflow.com/invocations/default/698b40b9-5776-4189-8fce-316febf5f82c?testReportRun=3&testReportShard=20&testReportAttempt=1#targets-Ly9wa2cva3Yva3ZzZXJ2ZXI6a3ZzZXJ2ZXJfdGVzdA==) on master @ [6d3f108b4de3ae24ebb1543ca3882144678f8fa2](https://github.com/cockroachdb/cockroach/commits/6d3f108b4de3ae24ebb1543ca3882144678f8fa2):


```
=== RUN   TestMergeQueue
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestMergeQueue2181856278
    test_log_scope.go:76: use -show-logs to present logs inline
    client_merge_test.go:4695: -- test log scope end --
test logs left over in: outputs.zip/logTestMergeQueue2181856278
--- FAIL: TestMergeQueue (234.43s)
=== RUN   TestMergeQueue/load-based-merging
    --- FAIL: TestMergeQueue/load-based-merging (226.87s)
=== RUN   TestMergeQueue/load-based-merging/switch-queries-per-second-to-cpu-per-second-prev-combined-below-threshold
    client_merge_test.go:4643: condition failed to evaluate within 3m45s: from client_merge_test.go:4320: ranges unexpectedly merged
        --- FAIL: TestMergeQueue/load-based-merging/switch-queries-per-second-to-cpu-per-second-prev-combined-below-threshold (226.04s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=3</code>
 - <code>shard=20</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMergeQueue.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>


Jira issue: CRDB-43297

--- a/pkg/kv/kvserver/client_merge_test.go
+++ b/pkg/kv/kvserver/client_merge_test.go
@@ -3171,7 +3171,6 @@ func TestMergeQueueWithExternalFiles(t *testing.T) {
 			store, err := s.GetStores().(*kvserver.Stores).GetStore(s.GetFirstStoreID())
 			require.NoError(t, err)
 
-			store.SetMergeQueueActive(true)
 			if skipExternal {
 				verifyUnmergedSoon(t, store, lhsDesc.StartKey, rhsDesc.StartKey)
 			} else {
@@ -4293,6 +4292,11 @@ func TestStoreRangeMergeDuringShutdown(t *testing.T) {
 
 func verifyMergedSoon(t *testing.T, store *kvserver.Store, lhsStartKey, rhsStartKey roachpb.RKey) {
 	t.Helper()
+	store.SetMergeQueueActive(true)
+	defer func() {
+		store.SetMergeQueueActive(false)
+		store.MustForceMergeScanAndProcess() // drain any merges that might already be queued
+	}()
 	testutils.SucceedsSoon(t, func() error {
 		store.MustForceMergeScanAndProcess()
 		repl := store.LookupReplica(rhsStartKey)
@@ -4310,6 +4314,11 @@ func verifyUnmergedSoon(
 	t *testing.T, store *kvserver.Store, lhsStartKey, rhsStartKey roachpb.RKey,
 ) {
 	t.Helper()
+	store.SetMergeQueueActive(true)
+	defer func() {
+		store.SetMergeQueueActive(false)
+		store.MustForceMergeScanAndProcess() // drain any merges that might already be queued
+	}()
 	testutils.SucceedsSoon(t, func() error {
 		store.MustForceMergeScanAndProcess()
 		repl := store.LookupReplica(rhsStartKey)
@@ -4344,9 +4353,6 @@ func TestMergeQueue(t *testing.T) {
 						WallClock:                 manualClock,
 						DefaultZoneConfigOverride: &zoneConfig,
 					},
-					Store: &kvserver.StoreTestingKnobs{
-						DisableScanner: true,
-					},
 				},
 			},
 		})
@@ -4354,11 +4360,6 @@ func TestMergeQueue(t *testing.T) {
 
 	conf := zoneConfig.AsSpanConfig()
 	store := tc.GetFirstStoreFromServer(t, 0)
-	// The cluster with manual replication disables the merge queue,
-	// so we need to re-enable.
-	_, err := tc.ServerConn(0).Exec(`SET CLUSTER SETTING kv.range_merge.queue.enabled = true`)
-	require.NoError(t, err)
-	store.SetMergeQueueActive(true)
 
 	split := func(t *testing.T, key roachpb.Key, expirationTime hlc.Timestamp) {
 		t.Helper()
@@ -4429,6 +4430,7 @@ func TestMergeQueue(t *testing.T) {
 			kvserver.SplitByLoadEnabled.Override(ctx, &s.ClusterSettings().SV, false)
 		}
 
+		store.SetMergeQueueActive(false)     // reset merge queue to inactive
 		store.MustForceMergeScanAndProcess() // drain any merges that might already be queued
 		split(t, rhsStartKey.AsRawKey(), hlc.Timestamp{} /* expirationTime */)
 	}
@@ -4818,7 +4820,8 @@ func TestMergeQueueSeesNonVoters(t *testing.T) {
 	}
 
 	var clusterArgs = base.TestClusterArgs{
-		// We dont want the replicate queue mucking with our test, so disable it.
+		// We don't want the replicate queue mucking with our test, so disable it.
+		// This also disables the merge queue, until it is manually enabled.
 		ReplicationMode: base.ReplicationManual,
 		ServerArgs: base.TestServerArgs{
 			Knobs: base.TestingKnobs{
@@ -4841,10 +4844,6 @@ func TestMergeQueueSeesNonVoters(t *testing.T) {
 
 			store, err := tc.Server(0).GetStores().(*kvserver.Stores).GetStore(1)
 			require.Nil(t, err)
-			// We're going to split the dummy range created above with an empty
-			// expiration time. Disable the merge queue before splitting so that the
-			// split ranges aren't immediately merged.
-			store.SetMergeQueueActive(false)
 			leftDesc, rightDesc := splitDummyRangeInTestCluster(
 				t, tc, dbName, "kv" /* tableName */, hlc.Timestamp{} /* splitExpirationTime */)
 
@@ -4887,7 +4886,6 @@ func TestMergeQueueSeesNonVoters(t *testing.T) {
 			tc.RemoveVotersOrFatal(t, rightDesc.StartKey.AsRawKey(), tc.Target(0))
 			rightDesc = tc.LookupRangeOrFatal(t, rightDesc.StartKey.AsRawKey())
 
-			store.SetMergeQueueActive(true)
 			verifyMergedSoon(t, store, leftDesc.StartKey, rightDesc.StartKey)
 		})
 	}
@@ -4909,7 +4907,8 @@ func TestMergeQueueWithSlowNonVoterSnaps(t *testing.T) {
 	ctx := context.Background()
 	var delaySnapshotTrap atomic.Value
 	var clusterArgs = base.TestClusterArgs{
-		// We dont want the replicate queue mucking with our test, so disable it.
+		// We don't want the replicate queue mucking with our test, so disable it.
+		// This also disables the merge queue, until it is manually enabled.
 		ReplicationMode: base.ReplicationManual,
 		ServerArgs: base.TestServerArgs{
 			Knobs: base.TestingKnobs{
@@ -4945,17 +4944,9 @@ func TestMergeQueueWithSlowNonVoterSnaps(t *testing.T) {
 	numNodes := 3
 	tc, _ := setupTestClusterWithDummyRange(t, clusterArgs, dbName, tableName, numNodes)
 	defer tc.Stopper().Stop(ctx)
-	// We're controlling merge queue operation via
-	// `store.SetMergeQueueActive`, so enable the cluster setting here.
-	_, err := tc.ServerConn(0).Exec(`SET CLUSTER SETTING kv.range_merge.queue.enabled=true`)
-	require.NoError(t, err)
 
 	store, err := tc.Server(0).GetStores().(*kvserver.Stores).GetStore(1)
 	require.Nil(t, err)
-	// We're going to split the dummy range created above with an empty
-	// expiration time. Disable the merge queue before splitting so that the
-	// split ranges aren't immediately merged.
-	store.SetMergeQueueActive(false)
 	leftDesc, rightDesc := splitDummyRangeInTestCluster(
 		t, tc, dbName, tableName, hlc.Timestamp{}, /* splitExpirationTime */
 	)
@@ -4972,7 +4963,6 @@ func TestMergeQueueWithSlowNonVoterSnaps(t *testing.T) {
 		time.Sleep(5 * time.Second)
 		return nil
 	})
-	store.SetMergeQueueActive(true)
 	verifyMergedSoon(t, store, leftDesc.StartKey, rightDesc.StartKey)
 }
 

